File: truth/frontend/src/requests/industries.ts
/**
 * Every call around the industries - the vocabulary the navigation and the assumption chips are built from.
 */

import type { Industry, IndustryDraft } from '@/models/industry'
import { client } from '@/requests/client'
import { readAll } from '@/requests/paging'

const INDUSTRIES_PATH = '/industry'

/**
 * Read one window of the industries.
 */
const listIndustries = async (offset: number, limit: number): Promise<Industry[]> => {
  const response = await client.get<Industry[]>(INDUSTRIES_PATH, { params: { offset, limit } })

  return response.data
}

/**
 * Read every registered industry.
 */
const readAllIndustries = async (onProgress?: (loaded: number) => void): Promise<Industry[]> =>
  readAll(listIndustries, onProgress)

/**
 * Read every industry together with how many assumptions name it.
 *
 * The counting is one pass over the register inside the service. It used to be done here, by walking every
 * assumption the client had read - which meant the count was wrong until the last of them had arrived, and
 * impossible once there were more of them than a browser could hold.
 */
const readIndustriesWithCounts = async (): Promise<Industry[]> =>
  readAll(async (offset, limit) => {
    const response = await client.get<Industry[]>(INDUSTRIES_PATH, {
      params: { offset, limit, with_counts: true },
    })

    return response.data
  })

/**
 * Read a single industry addressed by its identifier.
 */
const readIndustry = async (industryId: string): Promise<Industry> => {
  const response = await client.get<Industry>(`${INDUSTRIES_PATH}/${industryId}`)

  return response.data
}

/**
 * Read a single industry addressed by its name.
 */
const readIndustryByName = async (name: string): Promise<Industry> => {
  const response = await client.get<Industry>(`${INDUSTRIES_PATH}/name/${encodeURIComponent(name)}`)

  return response.data
}

/**
 * Register a new industry and hand back the identifier it was given.
 */
const createIndustry = async (draft: IndustryDraft): Promise<string> => {
  const response = await client.post<{ id: string }>(INDUSTRIES_PATH, draft)

  return response.data.id
}

export {
  createIndustry,
  listIndustries,
  readAllIndustries,
  readIndustriesWithCounts,
  readIndustry,
  readIndustryByName,
}

File: truth/backend/src/truth_service/api/industries.py
"""
The endpoints of the industries, which are the vocabulary the register is filed under.

:date: 2026-09-08
"""
# ----- IMPORTS ----- #

from fastapi import APIRouter, Query, status

from truth_service.api.pagination import LIMIT_QUERY, OFFSET_QUERY
from truth_service.dependencies import IndustryServiceDependency
from truth_service.models.industry import IndustryCreateRequest, IndustryResponse

# ----- CONSTS ----- #

ROUTER: APIRouter = APIRouter(prefix="/industry", tags=["industries"])

COUNTS_QUERY = Query(default=False, description="Whether each industry is answered with its assumption count")

# ----- FUNCTIONS ----- #


@ROUTER.get("", response_model=list[IndustryResponse])
async def list_industries(
    service: IndustryServiceDependency,
    offset: int = OFFSET_QUERY,
    limit: int = LIMIT_QUERY,
    with_counts: bool = COUNTS_QUERY,
) -> list[IndustryResponse]:
    """
    Read one window of the registered industries.

    :param service: Owner of the industries.
    :param offset: Amount of industries skipped before collecting.
    :param limit: Largest amount of industries that is returned.
    :param with_counts: Whether each industry is answered with how many assumptions name it.
    :return: The industries of the window.
    """
    return await service.list_industries(offset=offset, limit=limit, with_counts=with_counts)


@ROUTER.post("", response_model=IndustryResponse, status_code=status.HTTP_201_CREATED)
async def create_industry(request: IndustryCreateRequest, service: IndustryServiceDependency) -> IndustryResponse:
    """
    Register a new industry.

    :param request: Industry supplied by the caller.
    :param service: Owner of the industries.
    :return: The industry as it was stored, carrying the identifier it was given.
    """
    return await service.create_industry(request=request)


@ROUTER.get("/name/{name}", response_model=IndustryResponse)
async def read_industry_by_name(name: str, service: IndustryServiceDependency) -> IndustryResponse:
    """
    Read one industry addressed by its name.

    :param name: Name of the industry.
    :param service: Owner of the industries.
    :return: The industry.
    """
    return await service.get_industry_by_name(name=name)


@ROUTER.get("/{industry_id}", response_model=IndustryResponse)
async def read_industry(industry_id: str, service: IndustryServiceDependency) -> IndustryResponse:
    """
    Read one industry addressed by its identifier.

    :param industry_id: Identifier of the industry.
    :param service: Owner of the industries.
    :return: The industry.
    """
    return await service.get_industry(key=industry_id)
